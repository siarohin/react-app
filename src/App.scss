#root {
  display: grid;
  grid-template-rows: auto auto var(--app-spacing-xxxl);
  height: 100vh;
}
