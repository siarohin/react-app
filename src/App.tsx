import React, { Component, ReactElement } from "react";

import { Header } from "./components";

class App extends Component {
  public render(): ReactElement {
    return (
      <div className="app">
        <Header title="Hello World" />
      </div>
    );
  }
}

export default App;
