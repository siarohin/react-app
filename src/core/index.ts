export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
