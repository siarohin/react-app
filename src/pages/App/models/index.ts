export * from "./AppState";
export * from "./IAppState";
