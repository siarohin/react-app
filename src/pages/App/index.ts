import App from "./App";

export * from "./AppState";
export { App };
