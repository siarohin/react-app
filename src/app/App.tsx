import { hot } from "react-hot-loader";
import React from "react";
import { Route, Switch } from "react-router-dom";
import { Helmet } from "react-helmet";
import loadable from "@loadable/component";

import "../../public/assets/styles.scss";
import { Endpoints, MetaTitle, META_CHARSET, RouterPath } from "./core";
import { CircularProgress } from "./shared";
import { MoviesListErrorBoundary } from "./containers";

const fallback = <CircularProgress className="app__progress-bar" size="8rem" color="secondary" />;
const Home = loadable(() => import("./pages/Home"), { fallback });
const NotFound = loadable(() => import("./pages/NotFound"));

const App = () => {
  const currentYear: string = new Date().getFullYear().toString();

  return (
    <>
      <Helmet defaultTitle={`${MetaTitle.DefaultTitle} in ${currentYear}`} titleTemplate="%s - React movie">
        <meta charSet={META_CHARSET} />
        <link rel="canonical" href={`${Endpoints.ClientHost}`} />
      </Helmet>

      <div className="app">
        <MoviesListErrorBoundary>
          <Switch>
            <Route exact path="/" component={Home} />
            <Route exact path={`${RouterPath.Movie}/:id`} component={Home} />
            <Route path={RouterPath.Search} component={Home} />
            <Route path="*" component={NotFound} />
          </Switch>
        </MoviesListErrorBoundary>
      </div>
    </>
  );
};

export default hot(module)(App);
