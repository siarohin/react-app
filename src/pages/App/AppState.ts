import { MoviesModels } from "../../core";
import { IAppState } from "./IAppState";

/**
 * App state
 */
export const AppState: IAppState = {
  movieDialog: {
    dialogSettings: {
      title: "",
      values: {} as MoviesModels.IMovie
    },
    open: false
  }
};
