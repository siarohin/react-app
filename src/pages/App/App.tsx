import React, { ComponentType, ReactElement, useCallback, useEffect, useMemo, useState } from "react";
import { Redirect, Route, Switch } from "react-router-dom";
import { connect } from "react-redux";
import Notifications, { NotificationsState } from "react-notification-system-redux";

import "./App.scss";
import { Footer, NoMovies } from "../../components";
import {
  MainLayout,
  MoviesFilter,
  MoviesList,
  MoviesListErrorBoundary,
  Header,
  HeaderLayout,
  MovieDialog,
  MovieDetails
} from "../../containers";
import { DialogTitle, DialogAction, State, MoviesModels, UserPreferencesModels, RouterPath } from "../../core";
import { IDialogSettings } from "../../models";
import { getEnumKey } from "../../utils";
import { AppState } from "./models";

const App = ({
  notifications,
  dialogAction
}: {
  notifications: NotificationsState;
  dialogAction: UserPreferencesModels.IMovieAction;
}): ReactElement => {
  const [movieDialog, setMovieDialog] = useState(AppState.movieDialog);

  const handleCloseDialog = useCallback(
    (): void => setMovieDialog({ dialogSettings: (undefined as unknown) as IDialogSettings, open: false }),
    [movieDialog]
  );

  useEffect(() => {
    try {
      const actionType: string = getEnumKey(DialogAction, dialogAction?.action);
      const dialogSettings: IDialogSettings = {
        title: DialogTitle[actionType as keyof typeof DialogTitle],
        values: dialogAction?.movie as MoviesModels.IMovie
      };
      setMovieDialog({ dialogSettings, open: true });
    } catch {
      return;
    }
  }, [dialogAction]);

  const getDialog = useMemo((): ReactElement<typeof MovieDialog> | null => {
    const hasNotDialog: boolean = !movieDialog.dialogSettings;
    return hasNotDialog ? null : (
      <MovieDialog dialogSettings={movieDialog.dialogSettings} open={movieDialog.open} onClose={handleCloseDialog} />
    );
  }, [movieDialog]);

  return (
    <>
      <HeaderLayout>
        <Switch>
          <Route exact path={`${RouterPath.Movie}/:id`} component={MovieDetails} />
          <Route path="*" component={Header} />
        </Switch>
      </HeaderLayout>
      <MainLayout>
        <MoviesFilter />
        <Switch>
          <Route exact path={[`${RouterPath.Movie}/:id`, `${RouterPath.Search}/:search`, RouterPath.Search]}>
            <MoviesListErrorBoundary>
              <MoviesList />
            </MoviesListErrorBoundary>
          </Route>
          <Route exact path="/" component={NoMovies} />
          <Redirect from="*" to={RouterPath.NotFound} />
        </Switch>
        {getDialog}
      </MainLayout>
      <Footer />
      <Notifications notifications={notifications} />
    </>
  );
};

const mapStateToProps = (
  state: State
): { notifications: NotificationsState; dialogAction: UserPreferencesModels.IMovieAction } => ({
  notifications: state.notifications,
  dialogAction: state.userPreferences.dialogAction
});

export default connect(mapStateToProps)(App as ComponentType<any>);
