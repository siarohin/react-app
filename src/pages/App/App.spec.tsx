import React from "react";
import { Route, BrowserRouter as Router, Switch } from "react-router-dom";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import { State } from "../../core";
import App from "./App";

const mockStore = configureStore([]);

describe("Pages.App.App: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({
      movieList: {
        totalAmount: 3000
      },
      userPreferences: {
        genres: {
          all: ["Romantic"],
          selected: ""
        },
        sortingOptions: {
          options: ["Rating", "Release date"],
          selected: "Release date"
        },
        search: {
          selected: ""
        }
      },
      notifications: []
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <Router>
          <Switch>
            <Route path="/" component={App} />
          </Switch>
        </Router>
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
