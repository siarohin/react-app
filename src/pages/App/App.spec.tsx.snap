// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages.App.App:  should render component 1`] = `
Array [
  <div
    className="app-header-layout"
  >
    <div
      className="app-header-layout__logo"
    >
      <h1
        className="app-logo"
      >
        React movie
      </h1>
    </div>
    <div
      className="app-header__top-right"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text app-add-movie-button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
           Add movie
        </span>
      </button>
    </div>
    <div
      className="app-header__container"
    >
      <div
        className="app-search-panel"
      >
        <h3
          className="app-search-panel__name"
        >
          Find your movie
        </h3>
        <form
          className="app-search-panel__form"
          onSubmit={[Function]}
        >
          <div
            className="MuiFormControl-root MuiTextField-root app-search-panel__form-field"
          >
            <div
              className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiFilledInput-input"
                disabled={false}
                name="searchInput"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="What do you want to watch?"
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text app-search-panel__form-button"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Search
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-1 app-main-layout"
  >
    <div
      className="app-movies-filter"
    >
      <div
        className="app-movies-filter__tab-panel"
      >
        <div
          aria-label="text alignment"
          className="MuiToggleButtonGroup-root app-filter-buttons"
          role="group"
        >
          <button
            aria-pressed={true}
            className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal app-filter-buttons__toggle-button Mui-selected"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
            value="All"
          >
            <span
              className="MuiToggleButton-label"
            >
              All
            </span>
          </button>
          <button
            aria-pressed={false}
            className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal app-filter-buttons__toggle-button"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
            value="Romantic"
          >
            <span
              className="MuiToggleButton-label"
            >
              Romantic
            </span>
          </button>
        </div>
      </div>
      <div
        className="app-movies-filter__select-panel"
      >
        <div
          className="app-filter-select__prefix"
        >
          Sort by
        </div>
        <div
          className="MuiFormControl-root app-filter-select__control"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline app-filter-select__select MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              Rating
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value="Rating"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        className="app-movies-filter__count"
      >
        <span
          className="app-filter-count"
        >
          3000
        </span>
         movies found
      </div>
      <div
        className="app-movies-filter__pagination"
      />
      <div
        className="app-movies-filter__separator"
      />
    </div>
    <div
      className="app-no-movies"
    >
      No Movies found
    </div>
  </div>,
  <div
    className="app-footer"
  >
    React movie
  </div>,
  <div
    className="notifications-wrapper"
    style={Object {}}
  />,
]
`;
